### PR TITLE
Validate prisoner - Add logic to validate a prisoners current prison code.

### PR DESCRIPTION
**Pull request by dhirajshettymoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/exception/BookerAlreadyExistsException.kt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package uk.gov.justice.digital.hmpps.prison.visitbooker.registry.exceptions
+package uk.gov.justice.digital.hmpps.prison.visitbooker.registry.exception
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/exception/BookerAlreadyExistsException.kt:1]: renamed package for consistency with other projects.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+) {
+  private companion object {
+    private val LOG = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun validatePrisoner(bookerReference: String, prisonerId: String) {
+    LOG.info("Validate booking by booker - {} for prisoner - {}", bookerReference, prisonerId)
+
+    val permittedPrisoner = bookerDetailsService.getPermittedPrisoner(bookerReference, prisonerId).let { permittedPrisoner ->
+      PermittedPrisonerDto(
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt:25]: This Dto takes a PermittedPrisoner as a constructor value. Could you do this:

.let { permittedPrisonerEntity -> PermittedPrisonerDto(permittedPrisonerEntity) }

or even

val permittedPrisoner = PermittedPrisonerDto(bookerDetailsService.getPermittedPrisoner(bookerReference, prisonerId))

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  private fun getPrisonerValidationErrorCodes(bookerReference: String, permittedPrisoner: PermittedPrisonerDto, prisonerSearchPrisoner: PrisonerDto): List<PrisonerValidationErrorCodes> {
+    val errorCodes = mutableListOf<PrisonerValidationErrorCodes>()
+
+    if (permittedPrisoner.prisonCode != prisonerSearchPrisoner.prisonId) {
+      LOG.info("Prison code {} on booker registry for prisoner - {} and booker reference - {} does not match with prison code - {} on prisoner offender search", permittedPrisoner.prisonCode, permittedPrisoner.prisonerId, bookerReference, prisonerSearchPrisoner.prisonId)
+      if (hasPrisonerBeenReleased(prisonerSearchPrisoner)) {
+        errorCodes.add(PRISONER_RELEASED)
+      } else {
+        errorCodes.add(PRISONER_TRANSFERRED)
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt:51]: As mentioned in enum class, if this is another ticket, maybe remove the else {} block until we move to that ticket.

File: src/main/resources/db/migration/V1_4__add_prison_code_permitted_prisoner.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE permitted_prisoner ADD prison_code VARCHAR(3);
+UPDATE permitted_prisoner set prison_code = 'NA' where prison_code IS NULL;
```
Comment (adaviesMOJ) [inline src/main/resources/db/migration/V1_4__add_prison_code_permitted_prisoner.sql:2]: What prison is NA ?

**Review by adaviesMOJ — COMMENTED** [review]
Some really minor comments / questions. Really nice work!

File: build.gradle.kts [hunk 5]
```diff
@@ -19,6 +20,7 @@ repositories {
 }
 
 dependencies {
+  implementation("uk.gov.justice.service.hmpps:hmpps-kotlin-spring-boot-starter:1.0.8")
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply build.gradle.kts:23]: The hmpps-template has a more concise WebClientConfiguration [https://github.com/ministryofjustice/hmpps-template-kotlin/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/templatepackagename/config/WebClientConfiguration.kt] which I updated to. This needs the kotlin-lib which uses the WebClient as mentioned here - https://github.com/ministryofjustice/hmpps-kotlin-lib/blob/1bf38c08142996598c7ebc9b206186a5518ff854/readme-contents/WebClients.md

File: src/main/resources/db/migration/V1_4__add_prison_code_permitted_prisoner.sql (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/resources/db/migration/V1_4__add_prison_code_permitted_prisoner.sql:2]: setting it to NA for the timebeing before setting it to not nullable. Maybe updating to Foston Hall code makes more sense.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/controller/BookerDetailsController.kt [hunk 6]
```diff
@@ -1,28 +1,37 @@
...
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.config.ValidationErrorResponse
 import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.dto.PermittedPrisonerDto
 import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.dto.PermittedVisitorDto
 import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.service.BookerDetailsService
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.service.PrisonerValidationService
 
 const val PUBLIC_BOOKER_CONTROLLER_PATH: String = "/public/booker/{bookerReference}"
 const val PERMITTED_PRISONERS: String = "$PUBLIC_BOOKER_CONTROLLER_PATH/permitted/prisoners"
 const val PERMITTED_VISITORS: String = "$PERMITTED_PRISONERS/{prisonerId}/permitted/visitors"
+const val VALIDATE_PRISONER: String = "$PUBLIC_BOOKER_CONTROLLER_PATH/prisoner/{prisonerId}/validate"
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/controller/BookerDetailsController.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/controller/BookerDetailsController.kt [hunk 7]
```diff
@@ -116,4 +125,50 @@ class BookerDetailsController(
...
+    @PathVariable(value = "prisonerId", required = true)
+    @Parameter(
+      description = "Prisoner Id for that needs to be validated.",
+      example = "A12345DC",
+    )
+    @NotBlank
+    prisonerId: String,
+  ): ResponseEntity<String> {
+    prisonerValidationService.validatePrisoner(bookerReference, prisonerId)
+    return ResponseEntity.status(HttpStatus.OK).body(
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/controller/BookerDetailsController.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/dto/enums/PrisonerValidationErrorCodes.kt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.justice.digital.hmpps.prison.visitbooker.registry.dto.enums
+
+enum class PrisonerValidationErrorCodes(
+  val description: String,
+) {
+  PRISONER_RELEASED("Prisoner released"),
+  PRISONER_TRANSFERRED("Prisoner transferred"),
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/dto/enums/PrisonerValidationErrorCodes.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt:24]: this is giving an error as the PermittedPrisonerDto() constructor call tries to lazily load the visitors foe a prisoner. As discussed, not updated but will look at ways to improve the constructor call later

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/PrisonerValidationService.kt]: updated

**Review by adaviesMOJ — APPROVED** [review]

**Review by adaviesMOJ — APPROVED** [review]

**Pull request merged**